File: c++/secure_password_generator.cpp
/*
GitHub Username: Shivi2705
Aim: Generate secure random passwords with user-defined options
Date: October 4, 2025
*/

#include <iostream>
#include <string>
#include <vector>
#include <ctime>
#include <cstdlib>
#include <algorithm>

using namespace std;

// Function to generate a secure random password
string generatePassword(int length, bool useUpper, bool useLower, bool useDigits, bool useSpecial) {
    if(length <= 0) {
        throw invalid_argument("Password length must be greater than 0");
    }

    string upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    string lower = "abcdefghijklmnopqrstuvwxyz";
    string digits = "0123456789";
    string special = "!@#$%^&*()-_=+[]{}|;:,.<>?/";
    string allChars = "";

    if(useUpper) allChars += upper;
    if(useLower) allChars += lower;
    if(useDigits) allChars += digits;
    if(useSpecial) allChars += special;

    if(allChars.empty()) {
        throw invalid_argument("At least one character type must be selected");
    }

    vector<char> password;

    // Ensure at least one character from each selected type
    if(useUpper) password.push_back(upper[rand() % upper.size()]);
    if(useLower) password.push_back(lower[rand() % lower.size()]);
    if(useDigits) password.push_back(digits[rand() % digits.size()]);
    if(useSpecial) password.push_back(special[rand() % special.size()]);

    // Fill remaining length
    while(password.size() < static_cast<size_t>(length)) {
        password.push_back(allChars[rand() % allChars.size()]);
    }

    // Shuffle password to avoid predictable patterns
    random_shuffle(password.begin(), password.end());

    return string(password.begin(), password.end());
}

int main() {
    srand(static_cast<unsigned int>(time(0)));

    cout << "=== Secure Random Password Generator ===\n";
    int length;
    cout << "Enter password length: ";
    cin >> length;

    char choice;
    bool useUpper, useLower, useDigits, useSpecial;

    cout << "Include uppercase letters? (y/n): ";
    cin >> choice;
    useUpper = (choice == 'y' || choice == 'Y');

    cout << "Include lowercase letters? (y/n): ";
    cin >> choice;
    useLower = (choice == 'y' || choice == 'Y');

    cout << "Include numbers? (y/n): ";
    cin >> choice;
    useDigits = (choice == 'y' || choice == 'Y');

    cout << "Include special characters? (y/n): ";
    cin >> choice;
    useSpecial = (choice == 'y' || choice == 'Y');

    try {
        string password = generatePassword(length, useUpper, useLower, useDigits, useSpecial);
        cout << "\n🔐 Generated Password: " << password << "\n";
    } catch(const invalid_argument &e) {
        cout << "\n❌ Error: " << e.what() << "\n";
    }

    return 0;
}

File: Python/secure_password_generator.py
"""
GitHub Username: Shivi2705
Aim: Generate secure random passwords with user-defined options
Date: October 4, 2025
"""

import random
import string

def generate_password(length, use_upper=True, use_lower=True, use_digits=True, use_special=True):
    """Generate a secure random password based on user choices."""
    
    if length <= 0:
        raise ValueError("Password length must be greater than 0")
    
    # Character pools
    upper = string.ascii_uppercase if use_upper else ''
    lower = string.ascii_lowercase if use_lower else ''
    digits = string.digits if use_digits else ''
    special = "!@#$%^&*()-_=+[]{}|;:,.<>?/" if use_special else ''
    
    all_chars = upper + lower + digits + special
    if not all_chars:
        raise ValueError("At least one character type must be selected")
    
    # Ensure at least one character from each selected type
    password = []
    if use_upper:
        password.append(random.choice(upper))
    if use_lower:
        password.append(random.choice(lower))
    if use_digits:
        password.append(random.choice(digits))
    if use_special:
        password.append(random.choice(special))
    
    # Fill the remaining length
    while len(password) < length:
        password.append(random.choice(all_chars))
    
    # Shuffle to avoid predictable patterns
    random.shuffle(password)
    return ''.join(password)

def main():
    print("=== Secure Random Password Generator ===")
    
    length = int(input("Enter password length: "))
    
    use_upper = input("Include uppercase letters? (y/n): ").lower() == 'y'
    use_lower = input("Include lowercase letters? (y/n): ").lower() == 'y'
    use_digits = input("Include numbers? (y/n): ").lower() == 'y'
    use_special = input("Include special characters? (y/n): ").lower() == 'y'
    
    password = generate_password(length, use_upper, use_lower, use_digits, use_special)
    print(f"\n🔐 Generated Password: {password}")

if __name__ == "__main__":
    main()
